tce/src/applibs/LLVMBackend/TDGen.cc

```diff
@@ -3557,6 +3560,38 @@ TDGen::writeControlFlowInstrDefs(std::ostream& os) {
     writeCallDef(os);
 }
 
+/**
+ * Writes hwloop instructions and patterns
+ */
+void
+TDGen::writeHWLoopDef(std::ostream& os) {
+    os << "// Hardware loop instructions" << std::endl
+       << "let isTerminator=1 in {" << std::endl
+       << "  def HWLOOPii : InstTCE<(outs), (ins i32imm0:$rIter, "
+          "i32imm0:$rInstr), \"\", []>;"
+       << std::endl
+       << "  def HWLOOPri : InstTCE<(outs), (ins i32imm0:$rIter, "
+          "i32imm0:$rInstr), \"\", []>;"
+       << std::endl
+       << "}" << std::endl
+       << "def : Pat<(int_set_loop_iterations i32imm0:$rIter), (HWLOOPii "
+          "i32imm0:$rIter, -1)>;"
+       << std::endl
+       << "def : Pat<(int_set_loop_iterations R32IRegs:$rIter), (HWLOOPri "
+          "R32IRegs:$rIter, -1)>;"
+       << std::endl
+
+       << "let isTerminator = 1 in {" << std::endl
+       << "  def LJUMP : InstTCE<(outs)," << std::endl
+       << "      (ins i32imm0:$rS, brtarget:$dst), \"\"," << std::endl
+       << "      [(brcond (int_loop_decrement i32imm0:$rS), bb:$dst)]>;"
+       << std::endl
+       << "}" << std::endl;
+
+    opNames_["HWLOOPii"] = "hwloop";
+    opNames_["HWLOOPri"] = "hwloop";
+    opNames_["LJUMP"] = "PSEUDO";
```

Comment:
This is explicit loop-index decrement + jump operation inserted by LLVM's hwloop.
Since we expect our control-unit to handle this, we select to PESUDO.

However, this can be used for break out from the loop in future.